application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/LicenseRenewListener.java

```diff
@@ -0,0 +1,208 @@
+        Arrays.asList(new ExtensionInstalledEvent(), new ExtensionUpgradedEvent());
+
+    @Inject
+    private LicensedExtensionManager licensedExtensionManager;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private Provider<Licensor> licensorProvider;
+
+    @Inject
+    private LicenseUpdater licenseUpdater;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public List<Event> getEvents()
+    {
+        return EVENTS;
+    }
+
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        // Retrieve license updates from store.
+        licenseUpdater.getLicensesUpdates();
+
+        ExtensionEvent extensionEvent = (ExtensionEvent) event;
+        // Only treat top level licensed extensions. If it's not a top level extension, then its license it's handled
+        // by another extension's license.
+        if (licensedExtensionManager.getLicensedExtensions().contains(extensionEvent.getExtensionId())
+            && isMandatoryLicensedExtension(extensionEvent.getExtensionId()))
+        {
+            InstalledExtension installedExtension = (InstalledExtension) source;
+            if (installedExtension == null) {
+                return;
+            }
+
+            License license = licensorProvider.get().getLicense(installedExtension.getId());
+            boolean hasLicense = license != null && !License.UNLICENSED.equals(license);
+
+            if (event instanceof ExtensionUpgradedEvent && hasLicense) {
+                extensionUpgraded(installedExtension, data, extensionEvent, license);
+            } else if (event instanceof ExtensionInstalledEvent) {
+                extensionInstalled(installedExtension, extensionEvent, hasLicense);
+            }
+        }
+    }
+
+    private void extensionInstalled(InstalledExtension installedExtension, ExtensionEvent extensionEvent,
+        boolean hasLicense)
+    {
+        if (hasLicense) {
+            logger.debug(
+                "[{}] has been installed and it has a license associated to it. Check if there are license changes "
+                    + "for this new extension version.", extensionEvent.getExtensionId());
+
+            // Since we don't have information on the old installed version, we trigger a license renew and let store
+            // check if this license needs changes.
+            licenseUpdater.renewLicense(installedExtension.getId());
+        }
+    }
+
+    private void extensionUpgraded(InstalledExtension installedExtension, Object data, ExtensionEvent extensionEvent,
+        License license)
+    {
+        InstalledExtension prevInstalledExtension =
+            getPreviousInstalledExtension(data, installedExtension, extensionEvent.getNamespace());
+        // Stop if previous version is actually higher, because we don't want to regenerate a license downgrade.
```

Comment:
Might be good to explain why we don't want to regenerate the license on downgrade (it's not obvious).